pkg/schemacompat/schemacompat.go

```diff
@@ -0,0 +1,374 @@
+package schemacompat
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// EnsureStructuralSchemaCompatibility compares a new structural schema to an existing one, to ensure that the existing
+// schema is a sub-schema of the new schema. In other words that means that all the documents validated by the existing schema
+// will also be validated by the new schema, so that the new schema can be considered backward-compatible with the existing schema.
+// If it's not the case, errors are reported for each incompatible schema change.
+//
+// PLEASE NOTE that the implementation is incomplete (it's ongoing work), but still consistent:
+// if some Json Schema elements are changed and the comparison of this type of element is not yet implemented,
+// then an incompatible change error is triggered explaining that the compatison on this element type is not supported.
+// So there should never be any case when a schema is considered backward-compatible while in fact it is not.
+//
+// If the narrowExisting argument is true, then the LCD (Lowest-Common-Denominator) between existing schema and the new schema
+// is built (when possible), and returned if no incompatible change was detected (like type change, etc...).
+// If the narrowExisting argument is false, the existing schema is untouched and no LCD schema is calculated.
+//
+// In either case, when no errors are reported, it is ensured that either the existing schema or the calculated LCD
+// is a sub-schema of the new schema.
+func EnsureStructuralSchemaCompatibility(fldPath *field.Path, existing, new *apiextensionsv1.JSONSchemaProps, narrowExisting bool) (lcd *apiextensionsv1.JSONSchemaProps, errors utilerrors.Aggregate) {
```

Comment:
This should probably just return `error`, which `utilerrors.Aggregate` is one of. Then you can avoid returning `utilerrors.NewAggregate([]error{err})` when you only have a regular error, and can just `return err`.

Reply:
TBF the meaning of the returned value is really a *list of errors*. I don't see how hiding this behind a generic `error` interface would make it better.
In most cases you would have to cast the returned error back to an `errors.Aggregate` (provided you first know it is an `Aggregate`), in order to manage errors individually (such as adding the error messages in a list whose format you control).
So if it's acceptable for you, I'd rather keep explicitly declaring the `Aggregate` return type, since you can still assign it to an `error` variable if you really need to.

Reply:
Handling and reporting multiple errors in Go is unfortunately really painful :(

I think my main complaint might be with `utilerrors.Aggregate`, which seems to be mainly designed only to be used by k8s apimachinery, and doesn't seem to be intended for arbitrary external usage. Things in the code like [`In case of input error list contains nil`](https://github.com/kubernetes/apimachinery/blob/v0.21.2/pkg/util/errors/errors.go#L35) and [`This should never happen, really`](https://github.com/kubernetes/apimachinery/blob/v0.21.2/pkg/util/errors/errors.go#L71) lead me to believe the authors have a very specific usage in mind, and if we do append a nil error or create an `Aggregate` of zero errors, we might get unexpected bugs -- unexpected by us, and by the authors.

I wonder if using a package designed specifically to signal multiple errors might be better, something like [Uber's `multierr`](https://github.com/uber-go/multierr) or [Hashicorp's `go-multierror`](https://github.com/hashicorp/go-multierror). Both of these are a bit more consumer-friendly because they let methods return a standard `error` if they want, but they can also return a multi-error if callers want to crack open the error to filter or format internal errors.

If we expect to make this method available to arbitrary callers, I think thinking through the optimal experience will be worth it, even if it means a bit more work on our end.

I don't think it necessarily needs to block this PR, but we should stay open to changing the return type before proposing more people use this. We might also find more sharp corners as we use it more ourselves.

Reply:
Thanks for the more detailed answer. I agree we might have to think about it more deeply when about to make this method available to arbitrary users. OTOH the comparison is limited to Kubernetes structural schemas, and take `*apiextensionsv1.JSONSchemaProps` arguments, which is also completely related to Kubernetes.
So while the goal is finally to share / contribute this method outside KCP, I'm not sure it makes sense outside of the Kubernetes ecosystem.